app/Dockerfile

```diff
@@ -33,5 +61,73 @@ ENV HOST=0.0.0.0
 # Install application dependencies.
 # https://python-poetry.org/docs/basic-usage/#installing-dependencies
 RUN poetry install
+
 # Run the application.
 CMD ["poetry", "run", "python", "-m", "api"]
+
+#-------------------------------------------------------------------------------
+# Non-development build environment
+#-------------------------------------------------------------------------------
+FROM build as app-build
+
+# Runtime arguement for what user will the bulk of this stage run as
+ARG RUN_USER
+
+# In between ARG RUN_USER and USER ${RUN_USER}, the user is still root
+# If there is anything that needs to be ran as root, this is the spot
+
+# Change the user from root to ${RUN_USER}
+USER ${RUN_USER}
+
+# Set the application working directory.
+WORKDIR /app
+
+# Install dependencies
+# Use a project-local virtualenv to be easy to find
+ENV POETRY_VIRTUALENVS_IN_PROJECT true
+COPY . /app
+RUN poetry install --no-root --without dev
+
+# Install application dependencies
+# The name of the wheel file comes from pyproject.toml
+RUN poetry build --format wheel && poetry run pip install 'dist/template_application_flask-0.1.0-py3-none-any.whl'
+
+#-------------------------------------------------------------------------------
+# Production Application Image only
+#-------------------------------------------------------------------------------
+# Now copy over just the installed components into the final image
+FROM scratch as release
+
+# Runtime arguement for what user will the bulk of this stage run as
+ARG RUN_USER
+
+# In between ARG RUN_USER and USER ${RUN_USER}, the user is still root
+# If there is anything that needs to be ran as root, this is the spot
+
+# Change the user from root to ${RUN_USER}
+USER ${RUN_USER}
+
+# Any required files or directories will need to be explicitly copied to
+# the release image
+COPY --from=app-build /etc/passwd /etc/group /etc/
+COPY --from=app-build /app/.venv /app/.venv
+
+COPY --from=app-build /usr/share/ca-certificates/* /usr/share/ca-certificates/
+COPY --from=app-build /usr/local/bin/python* /usr/local/bin/
+
+# Include etc files that link to /usr/local/lib/
+COPY --from=app-build /etc/ld.so.conf /etc/ld.so.cache /etc/
+COPY --from=app-build /etc/ld.so.conf.d/ /etc/ld.so.conf.d/
+
+# Include required libs
+COPY --from=app-build /usr/lib/x86_64-linux-gnu/ /usr/lib/x86_64-linux-gnu/
+COPY --from=app-build /lib/* /lib/
+COPY --from=app-build /usr/local/lib/ /usr/local/lib/
+COPY --from=app-build /usr/local/include/python* /usr/local/include/
+COPY --from=app-build /lib64/* /lib64/
+COPY --from=app-build --chown=${RUN_USER} /tmp/ /tmp/
+
+ENV PATH="/app/.venv/bin:$PATH"
+
+# run the application
+CMD ["run-api"]
```

Comment:
I'll add a comment about it, but there are two things creates the directory for the app, and then copy it over. 
 - In `app-build`, we use a virtual environment and wheel to create the files in the working directory, `app`, named `.venv`. This directory contains the files needed for the app, and any non-python code that needs to be copied over, a la `openapi.yml`, would need to be explicitly listed in the `pyproject.toml` file in the `include` line at the top of the file
 - In `release`, the files in `app/.venv` is copied over from the line `COPY --from=app-build /app/.venv /app/.venv`. The python module (ie `requests`, 'logger`, etc) directories are required are copied over in a later command `COPY --from=app-build /usr/local/include/python* /usr/local/include/`

Once both of these are set, we tell the path that it can run from `app/.venv/bin` and it knows, from being built as a wheel, to run the scripts specified in `pyproject.toml` as actual commands like this

Reply:
I get it thanks!

btw do we actually need to include openapi.yml in the release build? I think the swagger docs just rely on the Marshmallow schemas and don't actually need the openapi.yml file. (the template uses apiflask which is a code-first approach to API rather than a spec-first approach)

Reply:
Good question, I am not 100% sure, but I'd think it might not be needed then? So, if I remove that file and run it locally, the code will generate the `/docs` endpoint?

Reply:
I think so

Reply:
Yeap, it worked. I commented out the code and added a blurb about when it is needed